apis/kueue/v1beta1/clusterqueue_types.go

```diff
@@ -0,0 +1,305 @@
+	Cohort string `json:"cohort,omitempty"`
+
+	// QueueingStrategy indicates the queueing strategy of the workloads
+	// across the queues in this ClusterQueue. This field is immutable.
+	// Current Supported Strategies:
+	//
+	// - StrictFIFO: workloads are ordered strictly by creation time.
+	// Older workloads that can't be admitted will block admitting newer
+	// workloads even if they fit available quota.
+	// - BestEffortFIFO: workloads are ordered by creation time,
+	// however older workloads that can't be admitted will not block
+	// admitting newer workloads that fit existing quota.
+	//
+	// +kubebuilder:default=BestEffortFIFO
+	// +kubebuilder:validation:Enum=StrictFIFO;BestEffortFIFO
+	QueueingStrategy QueueingStrategy `json:"queueingStrategy,omitempty"`
+
+	// namespaceSelector defines which namespaces are allowed to submit workloads to
+	// this clusterQueue. Beyond this basic support for policy, an policy agent like
+	// Gatekeeper should be used to enforce more advanced policies.
+	// Defaults to null which is a nothing selector (no namespaces eligible).
+	// If set to an empty selector `{}`, then all namespaces are eligible.
+	NamespaceSelector *metav1.LabelSelector `json:"namespaceSelector,omitempty"`
+
+	// preemption describes policies to preempt Workloads from this ClusterQueue
+	// or the ClusterQueue's cohort.
+	//
+	// Preemption can happen in two scenarios:
+	//
+	// - When a Workload fits within the min quota of the ClusterQueue, but the
+	//   quota is currently borrowed by other ClusterQueues in the cohort.
+	//   Preempting Workloads in other ClusterQueues allows this ClusterQueue to
+	//   reclaim its min quota.
+	// - When a Workload doesn't fit within the min quota of the ClusterQueue
+	//   and there are active Workloads with lower priority.
+	//
+	// The preemption algorithm tries to find a minimal set of Workloads to
+	// preempt to accomomdate the pending Workload, preempting Workloads with
+	// lower priority first.
+	Preemption *ClusterQueuePreemption `json:"preemption,omitempty"`
+}
+
+type QueueingStrategy string
+
+const (
+	// StrictFIFO means that workloads are ordered strictly by creation time.
+	// Older workloads that can't be admitted will block admitting newer
+	// workloads even if they fit available quota.
+	StrictFIFO QueueingStrategy = "StrictFIFO"
+
+	// BestEffortFIFO means that workloads are ordered by creation time,
+	// however older workloads that can't be admitted will not block
+	// admitting newer workloads that fit existing quota.
+	BestEffortFIFO QueueingStrategy = "BestEffortFIFO"
+)
+
+type ResourceGroup struct {
+	// coveredResources is the list of resources covered by the flavors in this
+	// group.
+	// Examples: cpu, memory, vendor.com/gpu.
+	// The list cannot be empty and it can contain up to 16 resources.
+	// +kubebuilder:validation:MinItems=1
+	// +kubebuilder:validation:MaxItems=16
+	CoveredResources []corev1.ResourceName `json:"coveredResources"`
```

Comment:
I read the design docs https://docs.google.com/document/d/1Uu4hfGxux4Wh_laqZMLxXdEVdty06Sb2DwB035hj700/edit?resourcekey=0-b7mU7mGPCkEfhjyYDsXOBg#heading=h.dkobprho4lxe and I got the intension about adding a slice of resource names as a visual aid, just want to make sure whether we all agree on this for it will increase the burden of users whether reading or maintaining, bit or more. cc @ahg-g

Reply:
In addition to being a visual aid, it makes the API more explicit, which is a good principle.

But I agree that it might add burden, specially for resource groups with a single resource. But, on the other hand, ClusterQueues should be APIs that don't change often.